image/oci_test.go

```diff
@@ -0,0 +1,353 @@
+			ManifestMIMEType: manifest.DockerV2Schema2MediaType,
+		}))
+	}
+}
+
+// oci1ImageSource is plausible enough for schema conversions in manifestOCI1.UpdatedImage() to work.
+type oci1ImageSource struct {
+	configBlobImageSource
+	ref reference.Named
+}
+
+func (OCIis *oci1ImageSource) Reference() types.ImageReference {
+	return refImageReferenceMock{OCIis.ref}
+}
+
+func newOCI1ImageSource(t *testing.T, dockerRef string) *oci1ImageSource {
+	realConfigJSON, err := ioutil.ReadFile("fixtures/oci1-config.json")
+	require.NoError(t, err)
+
+	ref, err := reference.ParseNamed(dockerRef)
+	require.NoError(t, err)
+
+	return &oci1ImageSource{
+		configBlobImageSource: configBlobImageSource{
+			f: func(digest string) (io.ReadCloser, int64, error) {
+				return ioutil.NopCloser(bytes.NewReader(realConfigJSON)), int64(len(realConfigJSON)), nil
+			},
+		},
+		ref: ref,
+	}
+}
+
+func TestManifestOCI1UpdatedImage(t *testing.T) {
+	originalSrc := newOCI1ImageSource(t, "httpd:latest")
+	original := manifestOCI1FromFixture(t, originalSrc, "oci1.json")
+
+	// LayerInfos:
+	layerInfos := append(original.LayerInfos()[1:], original.LayerInfos()[0])
+	res, err := original.UpdatedImage(types.ManifestUpdateOptions{
+		LayerInfos: layerInfos,
+	})
+	require.NoError(t, err)
+	assert.Equal(t, layerInfos, res.LayerInfos())
+	_, err = original.UpdatedImage(types.ManifestUpdateOptions{
+		LayerInfos: append(layerInfos, layerInfos[0]),
+	})
+	assert.Error(t, err)
+
+	// ManifestMIMEType:
+	// Only smoke-test the valid conversions, detailed tests are below. (This also verifies that “original” is not affected.)
+	for _, mime := range []string{
+		manifest.DockerV2Schema2MediaType,
+	} {
+		_, err = original.UpdatedImage(types.ManifestUpdateOptions{
+			ManifestMIMEType: mime,
+			InformationOnly: types.ManifestUpdateInformation{
+				Destination: &memoryImageDest{ref: originalSrc.ref},
+			},
+		})
+		assert.NoError(t, err, mime)
+	}
+	for _, mime := range []string{
+		imgspecv1.MediaTypeImageManifest, // This indicates a confused caller, not a no-op.
+		"this is invalid",
+	} {
+		_, err = original.UpdatedImage(types.ManifestUpdateOptions{
+			ManifestMIMEType: mime,
+		})
+		assert.Error(t, err, mime)
+	}
+
+	// m hasn’t been changed:
+	m2 := manifestOCI1FromFixture(t, originalSrc, "oci1.json")
+	typedOriginal, ok := original.(*manifestOCI1)
+	require.True(t, ok)
+	typedM2, ok := m2.(*manifestOCI1)
+	require.True(t, ok)
+	assert.Equal(t, *typedM2, *typedOriginal)
+}
+
+func TestConvertToManifestSchema2(t *testing.T) {
+	originalSrc := newOCI1ImageSource(t, "httpd-copy:latest")
+	original := manifestOCI1FromFixture(t, originalSrc, "oci1.json")
+	memoryDest := &memoryImageDest{ref: originalSrc.ref}
+	res, err := original.UpdatedImage(types.ManifestUpdateOptions{
+		ManifestMIMEType: manifest.DockerV2Schema2MediaType,
+		InformationOnly: types.ManifestUpdateInformation{
+			Destination: memoryDest,
```

Comment:
This is unused, so it would be possible (indeed a bit desirable) not to provide that information.